src/Grpc.AspNetCore.Server/Internal/PipeExtensions.cs

```diff
@@ -194,9 +201,16 @@ public static async ValueTask<T> ReadSingleMessageAsync<T>(this PipeReader input
                                 // Finished and the complete message has arrived
                                 GrpcServerLog.DeserializingMessage(logger, (int)data.GetValueOrDefault().Length, typeof(T));
 
-                                serverCallContext.DeserializationContext.SetPayload(data);
-                                request = deserializer(serverCallContext.DeserializationContext);
-                                serverCallContext.DeserializationContext.SetPayload(null);
+                                if (marshaller is BufferMarshaller<T> bufferMarshaller)
```

Comment:
this is not necessary. 
If the generated code provides a marshaller that provides contextualSerializer and contextual deserializer, you should not need to differentiate between the buffer and non-buffer codepaths. 

https://github.com/grpc/grpc/blob/d18b52f5db44b1bfae42a08b3622a0d9b3688fa2/src/csharp/Grpc.Core.Api/Marshaller.cs#L121

See and example here: https://github.com/grpc/grpc/blob/d18b52f5db44b1bfae42a08b3622a0d9b3688fa2/src/csharp/Grpc.Core.Tests/ContextualMarshallerTest.cs#L45